Explicitly delete c-ares fd's from pollset set after c-ares is done

apolcyn opened the pull request:
In all of the existing tests for c-ares right now, the driving pollset set is quickly and completely unref'd and destroyed as soon as the test is over (by the channel). Google default credentials introduces a special case in that the google default credentials object is created as a global and potentially left around for a long time -  it's user receives a [second ref of it](https://github.com/grpc/grpc/blob/master/src/core/lib/security/credentials/google_default/google_default_credentials.cc#L247). It appears that the OAuth fetcher within google default credentials effectively [refs that driving pollset set](https://github.com/grpc/grpc/blob/master/src/core/lib/security/credentials/oauth2/oauth2_credentials.cc#L284) but then doesn't let that ref go until [it's destroyed](https://github.com/grpc/grpc/blob/master/src/core/lib/security/credentials/oauth2/oauth2_credentials.cc#L113), and as a result when c-ares is used with google default creds, the fd's created by c-ares are added to driving pollset set are not unref'd in a timely manner.

This test is a little messy but it seems to be a minimal way of catching the earlier bug.

(as a slightly separate issue, I'm having a hard time telling when `g_default_credentials` is fully unref'd - it <i>seems</i> to need `grpc_flush_cached_google_default_credentials` to be called?).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/grpc/grpc/15828)
<!-- Reviewable:end -->

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
  +0.0%     +99 [None]                                                                              +339  +0.0%
  +3.3%     +48 src/core/lib/iomgr/iomgr.cc                                                          +48  +3.3%
      [NEW]     +41 grpc_iomgr_count_objects_for_testing                                                 +41  [NEW]
       +15%      +7 [Unmapped]                                                                            +7   +15%
  +3.6%     +13 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver_posi     +13  +3.6%
       +13%     +13 grpc_core::GrpcPolledFdPosix::~GrpcPolledFdPosix                                     +13   +13%
      +4.0%      +4 grpc_core::NewGrpcPolledFdLocked                                                      +4  +4.0%

  +0.0%    +160 TOTAL                                                                               +400  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] Performance differences noted:
Benchmark                                                                cpu_time    real_time
-----------------------------------------------------------------------  ----------  -----------
BM_PumpStreamServerToClient<InProcess>/262144                            +7%         +7%
BM_StreamingPingPong<InProcess, NoOpMutator, NoOpMutator>/262144/1       -4%         -4%
BM_StreamingPingPongMsgs<InProcess, NoOpMutator, NoOpMutator>/32768      -4%         -4%
BM_StreamingPingPongMsgs<MinInProcess, NoOpMutator, NoOpMutator>/262144  -4%         -4%
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/0/2097152          +4%         +4%
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/0/262144           +5%         +5%
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/262144/262144      +8%         +8%
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/0/262144        +4%         +4%
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/262144/262144   +9%         +9%
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
  +0.0%     +99 [None]                                                                              +339  +0.0%
  +3.3%     +48 src/core/lib/iomgr/iomgr.cc                                                          +48  +3.3%
      [NEW]     +41 grpc_iomgr_count_objects_for_testing                                                 +41  [NEW]
       +15%      +7 [Unmapped]                                                                            +7   +15%
  +3.6%     +13 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver_posi     +13  +3.6%
       +13%     +13 grpc_core::GrpcPolledFdPosix::~GrpcPolledFdPosix                                     +13   +13%
      +4.0%      +4 grpc_core::NewGrpcPolledFdLocked                                                      +4  +4.0%

  +0.0%    +160 TOTAL                                                                               +400  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

markdroth commented on src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver_posix.cc:
```diff
@@ -41,14 +41,15 @@ namespace grpc_core {
 
   ~GrpcPolledFdPosix() {
     gpr_free(name_);
+    grpc_pollset_set_del_fd(driver_pollset_set_, fd_);
```
Do we have some guarantee that this pollset_set will still exist when this fd is destroyed?

apolcyn replied:
Good point, I don't think so. Canged the resolver to create and destroy it's own pollset set.

apolcyn replied:
Yeah I agree, and my reason for thinking this is the right approach is by looking at ownership of the fd's: the `AresDnsResolver`'s pollset set is destroyed upon `AresDnsResolver` [destruction](https://github.com/grpc/grpc/blob/master/src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc#L166). The entire c-ares query, which creates the event driver and fd's, is protected by a [ref](https://github.com/grpc/grpc/blob/master/src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc#L411) that is released upon the query's completion. That completion begins [its scheduling](https://github.com/grpc/grpc/blob/master/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc#L159) from a callback that's invoked during a call to `ares_process` on an fd though, and then remaining fd's in the driver's list are scheduled to be shut-down/cancelled/destroyed. So it would appear that the original `AresDnsResolver` might have been unref'd by the time that a <i>second</i> driver fd is destroyed. It seems tricky to test for this though...

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
  +0.0%    +106 [None]                                                                              +394  +0.0%
  +3.3%     +48 src/core/lib/iomgr/iomgr.cc                                                          +48  +3.3%
      [NEW]     +41 grpc_iomgr_count_objects_for_testing                                                 +41  [NEW]
       +15%      +7 [Unmapped]                                                                            +7   +15%
   +11%     +38 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver_posi     +38   +11%
       +23%     +22 grpc_core::GrpcPolledFdPosix::~GrpcPolledFdPosix                                     +22   +23%
       +22%     +22 grpc_core::NewGrpcPolledFdLocked                                                     +22   +22%

  +0.0%    +192 TOTAL                                                                               +480  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] Performance differences noted:
Benchmark                                                                              cpu_time    real_time
-------------------------------------------------------------------------------------  ----------  -----------
BM_StreamingPingPong<MinInProcess, NoOpMutator, NoOpMutator>/262144/1                  -4%         -4%
BM_StreamingPingPongMsgs<InProcess, NoOpMutator, NoOpMutator>/262144                   -4%         -4%
BM_StreamingPingPongMsgs<MinInProcess, NoOpMutator, NoOpMutator>/32768                 -6%         -6%
BM_StreamingPingPongWithCoalescingApi<InProcess, NoOpMutator, NoOpMutator>/262144/1/0  -4%         -4%
BM_StreamingPingPongWithCoalescingApi<InProcess, NoOpMutator, NoOpMutator>/262144/2/0  -6%         -6%
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/0/262144                         +6%         +6%
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
  +0.0%     +99 [None]                                                                              +331  +0.0%
  +3.3%     +48 src/core/lib/iomgr/iomgr.cc                                                          +48  +3.3%
      [NEW]     +41 grpc_iomgr_count_objects_for_testing                                                 +41  [NEW]
       +15%      +7 [Unmapped]                                                                            +7   +15%
  +3.6%     +13 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver_posi     +13  +3.6%
       +13%     +13 grpc_core::GrpcPolledFdPosix::~GrpcPolledFdPosix                                     +13   +13%
      +5.0%      +5 grpc_core::NewGrpcPolledFdLocked                                                      +5  +5.0%

  +0.0%    +160 TOTAL                                                                               +392  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
  +0.0%    +106 [None]                                                                              +394  +0.0%
  +3.3%     +48 src/core/lib/iomgr/iomgr.cc                                                          +48  +3.3%
      [NEW]     +41 grpc_iomgr_count_objects_for_testing                                                 +41  [NEW]
       +15%      +7 [Unmapped]                                                                            +7   +15%
   +11%     +38 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver_posi     +38   +11%
       +23%     +22 grpc_core::GrpcPolledFdPosix::~GrpcPolledFdPosix                                     +22   +23%
       +22%     +22 grpc_core::NewGrpcPolledFdLocked                                                     +22   +22%

  +0.0%    +192 TOTAL                                                                               +480  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

markdroth commented:
It looks like this approach involves creating a new pollset_set for each individual fd, and there may be several fds per request.  Please check with @sreecha about whether this is reasonable -- in particular, I'm not sure how much overhead is involved in creating and destroying pollset_sets.  I also seem to recall that there are some weird, counter-intuitive semantics surrounding removing pollset_sets that have been added to other pollset_sets.

If there's significant overhead, then an alternative might be to create a pollset_set in the ev driver and use that for all of the fds.  This would at least reduce the overhead to one pollset_set per request, rather than one per fd.  And I think it would eliminate the lifetime issues, since IIRC (but correct me if I'm wrong) the ev driver lives until after all of the fds are destroyed.

apolcyn commented:
Just a note that in general, I think that in a common case (certainly this is the case in our tests), the ares event drivers are usually going to be one-to-one with a single fd (since more fd's are only created if a new DNS server is contact and/or TCP is used).

> If there's significant overhead, then an alternative might be to create a pollset_set in the ev driver and use that for all of the fds. This would at least reduce the overhead to one pollset_set per request, rather than one per fd. And I think it would eliminate the lifetime issues, since IIRC (but correct me if I'm wrong) the ev driver lives until after all of the fds are destroyed.

Ack this would sound good to me too.

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
  +0.0%     +83 [None]                                                                              +419  +0.0%
  +2.1%     +48 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver.cc       +48  +2.1%
      +6.5%     +24 grpc_ares_ev_driver_create_locked                                                    +24  +6.5%
       +33%     +16 [Unmapped]                                                                           +16   +33%
      +4.7%      +8 grpc_ares_ev_driver_unref                                                             +8  +4.7%
  +3.3%     +48 src/core/lib/iomgr/iomgr.cc                                                          +48  +3.3%
      [NEW]     +41 grpc_iomgr_count_objects_for_testing                                                 +41  [NEW]
       +15%      +7 [Unmapped]                                                                            +7   +15%
  +3.6%     +13 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver_posi     +13  +3.6%
       +13%     +13 grpc_core::GrpcPolledFdPosix::~GrpcPolledFdPosix                                     +13   +13%
      +4.0%      +4 grpc_core::NewGrpcPolledFdLocked                                                      +4  +4.0%

  +0.0%    +192 TOTAL                                                                               +528  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

apolcyn commented:
one more option here might be to restructure things so that [the query completion](https://github.com/grpc/grpc/blob/master/src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc#L419) callback isn't invoked until the c-ares event driver (and its fd's) are destroyed, but I think that would need a more involved change.

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
  +0.0%    +243 [None]                                                                              +755  +0.0%
  +3.3%     +48 src/core/lib/iomgr/iomgr.cc                                                          +48  +3.3%
      [NEW]     +41 grpc_iomgr_count_objects_for_testing                                                 +41  [NEW]
       +17%      +7 [Unmapped]                                                                            +7   +17%
  +1.4%     +32 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver.cc       +32  +1.4%
      [NEW]    +380 grpc_ares_ev_driver_create_locked                                                   +380  [NEW]
       +24%     +12 [Unmapped]                                                                           +12   +24%
      [NEW]      +9 grpc_ares_ev_driver_on_queries_complete_locked                                        +9  [NEW]
      +4.7%      +8 grpc_ares_ev_driver_unref                                                             +8  +4.7%
  +0.3%     +16 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc         +16  +0.3%
      [NEW]    +236 log_address_sorting_list(grpc_lb_addresses*, char const*) [clone .isra.1]           +236  [NEW]
      [NEW]     +73 grpc_ares_complete_request_locked                                                    +73  [NEW]
      +2.7%     +24 on_hostbyname_done_locked                                                            +24  +2.7%
       +15%     +18 [Unmapped]                                                                           +18   +15%
      +1.5%      +8 on_srv_query_done_locked                                                              +8  +1.5%
      +0.4%      +6 grpc_dns_lookup_ares_locked_impl                                                      +6  +0.4%
  +3.6%     +13 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver_posi     +13  +3.6%
       +13%     +13 grpc_core::GrpcPolledFdPosix::~GrpcPolledFdPosix                                     +13   +13%
      +4.0%      +4 grpc_core::NewGrpcPolledFdLocked                                                      +4  +4.0%

  +0.0%    +352 TOTAL                                                                               +864  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
  +0.0%    +243 [None]                                                                              +739  +0.0%
  +3.3%     +48 src/core/lib/iomgr/iomgr.cc                                                          +48  +3.3%
      [NEW]     +41 grpc_iomgr_count_objects_for_testing                                                 +41  [NEW]
       +17%      +7 [Unmapped]                                                                            +7   +17%
  +1.4%     +32 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver.cc       +32  +1.4%
      [NEW]    +380 grpc_ares_ev_driver_create_locked                                                   +380  [NEW]
       +24%     +12 [Unmapped]                                                                           +12   +24%
      [NEW]      +9 grpc_ares_ev_driver_on_queries_complete_locked                                        +9  [NEW]
      +4.7%      +8 grpc_ares_ev_driver_unref                                                             +8  +4.7%
  +0.3%     +16 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc         +16  +0.3%
      [NEW]    +236 log_address_sorting_list(grpc_lb_addresses*, char const*) [clone .isra.1]           +236  [NEW]
      [NEW]     +73 grpc_ares_complete_request_locked                                                    +73  [NEW]
      +2.7%     +24 on_hostbyname_done_locked                                                            +24  +2.7%
       +15%     +18 [Unmapped]                                                                           +18   +15%
      +1.5%      +8 on_srv_query_done_locked                                                              +8  +1.5%
      +0.4%      +6 grpc_dns_lookup_ares_locked_impl                                                      +6  +0.4%
  +3.6%     +13 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver_posi     +13  +3.6%
       +13%     +13 grpc_core::GrpcPolledFdPosix::~GrpcPolledFdPosix                                     +13   +13%
      +4.0%      +4 grpc_core::NewGrpcPolledFdLocked                                                      +4  +4.0%

  +0.0%    +352 TOTAL                                                                               +848  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

markdroth commented:
Looking at the new integration between the request and EV driver objects, I think this is really screaming out for the two to be combined into a single object.  I won't insist upon doing that in this PR, but I think we should do that before C++-ifying this code, because it will be really hard to represent the current relationship between these objects using any of our C++ ref-counting classes.

---

Reviewed 2 of 4 files at r1, 1 of 3 files at r5, 1 of 2 files at r6, 4 of 4 files at r7.
Review status: all files reviewed, 3 unresolved discussions (waiting on @apolcyn, @dgquintas, @yihuazhang, @sreecha, and @AspirinSJL)

---

*[src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc, line 147 at r7](https://reviewable.io/reviews/grpc/grpc/15828#-LFs9fBs2mwqeDTrUFK0:-LFs9fBs2mwqeDTrUFK1:blnjl5d) ([raw file](https://github.com/grpc/grpc/blob/4b022bcd0d3c51eadf3bd8766af30c793e3cd79b/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc#L147)):*
> ```cpp
> }
> 
> static void grpc_ares_request_increment_pending_queries_locked(
> ```

I think these functions could still be called `ref` and `unref`.  The fact that the underlying type is a `size_t` instead of a `gpr_refcount` is an implementation detail; the external functionality is still the same here.

---

*[src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc, line 416 at r7](https://reviewable.io/reviews/grpc/grpc/15828#-LFsA9qL7pnkjNDv0co2:-LFsA9qL7pnkjNDv0co3:bodtri8) ([raw file](https://github.com/grpc/grpc/blob/4b022bcd0d3c51eadf3bd8766af30c793e3cd79b/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc#L416)):*
> ```cpp
>   r->error = GRPC_ERROR_NONE;
>   r->pending_queries = 0;
>   error = grpc_ares_ev_driver_create_locked(&r->ev_driver, interested_parties,
> ```

Why was this moved?

This seems incorrect to me.  The `goto error_cleanup` results in us returning `nullptr`, which means that we're not returning `r`, but we are not deallocating it anywhere.  It looks like this would result in a memory leak upon error.

---

*[src/core/lib/iomgr/iomgr.h, line 39 at r7](https://reviewable.io/reviews/grpc/grpc/15828#-LFs364r7K29rk8sdRFK:-LFs364r7K29rk8sdRFL:brxq3ey) ([raw file](https://github.com/grpc/grpc/blob/4b022bcd0d3c51eadf3bd8766af30c793e3cd79b/src/core/lib/iomgr/iomgr.h#L39)):*
> ```objc
> 
> /* Exposed only for testing */
> size_t grpc_iomgr_count_objects_for_testing();
> ```

Instead of exposing this, would it be sufficient to run the test with the environment variable `GRPC_ABORT_ON_LEAKS=1`?

---


*Comments from [Reviewable](https://reviewable.io/reviews/grpc/grpc/15828#-:-LFsAgtB6leIRJBDrdxE:b-vhu7a1)*
<!-- Sent from Reviewable.io -->

grpc-testing commented:
```
[trickle] No significant performance differences
```

apolcyn commented:
Review status: all files reviewed, 3 unresolved discussions (waiting on @apolcyn, @markdroth, @dgquintas, @yihuazhang, @sreecha, and @AspirinSJL)

---

*[src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc, line 147 at r7](https://reviewable.io/reviews/grpc/grpc/15828#-LFs9fBs2mwqeDTrUFK0:-LFuVZwyA0I1TQrLklvS:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/4b022bcd0d3c51eadf3bd8766af30c793e3cd79b/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc#L147)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

I think these functions could still be called `ref` and `unref`.  The fact that the underlying type is a `size_t` instead of a `gpr_refcount` is an implementation detail; the external functionality is still the same here.
</blockquote></details>

Done.

---

*[src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc, line 416 at r7](https://reviewable.io/reviews/grpc/grpc/15828#-LFsA9qL7pnkjNDv0co2:-LFuXr2L6rxKxRNbM362:bf473yk) ([raw file](https://github.com/grpc/grpc/blob/4b022bcd0d3c51eadf3bd8766af30c793e3cd79b/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc#L416)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Why was this moved?

This seems incorrect to me.  The `goto error_cleanup` results in us returning `nullptr`, which means that we're not returning `r`, but we are not deallocating it anywhere.  It looks like this would result in a memory leak upon error.
</blockquote></details>

I needed to move `grpc_ares_ev_driver_create_locked` below the creation of the request object only because this PR changes `grpc_ares_ev_driver_create_locked` to take the request object as a param.
But good point, I've added a `gpr_free(r)` into the `error_cleanup` label  for this.

---

*[src/core/lib/iomgr/iomgr.h, line 39 at r7](https://reviewable.io/reviews/grpc/grpc/15828#-LFs364r7K29rk8sdRFK:-LFuYdwpBVJXiXaVN2XZ:bkalv83) ([raw file](https://github.com/grpc/grpc/blob/4b022bcd0d3c51eadf3bd8766af30c793e3cd79b/src/core/lib/iomgr/iomgr.h#L39)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Instead of exposing this, would it be sufficient to run the test with the environment variable `GRPC_ABORT_ON_LEAKS=1`?
</blockquote></details>

Unfortunately I don't think it can help here. As far as I can see, `GRPC_ABORT_ON_LEAKS` can only  catch the case in which an fd is <i>leaked</i> even after all grpc objects are destroyed and `grpc_shutdown` is called. The new test in this PR is concerned not so much about <i>if</i> c-ares fd's are cleaned up, but instead <i>when</i> (currently, c-ares fd's are getting destroyed and cleaned up properly after all pollset sets that they've been added to are destroyed - but we need to make sure to destroy and free them as soon as they are no longer in use, otherwise we're potentially hanging on to them over the process lifetime).

---


*Comments from [Reviewable](https://reviewable.io/reviews/grpc/grpc/15828)*
<!-- Sent from Reviewable.io -->

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

markdroth commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed, 1 unresolved discussion (waiting on @markdroth, @dgquintas, @yihuazhang, @sreecha, and @AspirinSJL)

---

*[src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc, line 416 at r7](https://reviewable.io/reviews/grpc/grpc/15828#-LFsA9qL7pnkjNDv0co2:-LFwRf92-C75b4IYxgo-:b-zbw61o) ([raw file](https://github.com/grpc/grpc/blob/4b022bcd0d3c51eadf3bd8766af30c793e3cd79b/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc#L416)):*
<details><summary><i>Previously, apolcyn wrote…</i></summary><blockquote>

I needed to move `grpc_ares_ev_driver_create_locked` below the creation of the request object only because this PR changes `grpc_ares_ev_driver_create_locked` to take the request object as a param.
But good point, I've added a `gpr_free(r)` into the `error_cleanup` label  for this.
</blockquote></details>

I think there's still a bug here.  There are a bunch of other places in this function where we run `gpr_free(r)` and then `goto error_cleanup`.  So adding a `gpr_free(r)` after `error_cleanup:` means that we will free it twice in those cases.

I suggest removing the ocurrences of `gpr_free(r)` that occur right before `goto error_cleanup`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/grpc/grpc/15828)*
<!-- Sent from Reviewable.io -->

apolcyn commented:
Ack on combining the request and driver objects into one. I'll leave that for a separate PR if though if sounds good

---

Review status: all files reviewed, 1 unresolved discussion (waiting on @apolcyn, @dgquintas, @yihuazhang, @sreecha, and @AspirinSJL)

---

*[src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc, line 416 at r7](https://reviewable.io/reviews/grpc/grpc/15828#-LFsA9qL7pnkjNDv0co2:-LFwxZo7Ba8k33UtmFYV:bgbwfyn) ([raw file](https://github.com/grpc/grpc/blob/4b022bcd0d3c51eadf3bd8766af30c793e3cd79b/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc#L416)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

I think there's still a bug here.  There are a bunch of other places in this function where we run `gpr_free(r)` and then `goto error_cleanup`.  So adding a `gpr_free(r)` after `error_cleanup:` means that we will free it twice in those cases.

I suggest removing the ocurrences of `gpr_free(r)` that occur right before `goto error_cleanup`.
</blockquote></details>

Whoops, done

---


*Comments from [Reviewable](https://reviewable.io/reviews/grpc/grpc/15828#-:-LFwxgnl4n0DUG8miWpE:bd8pmsn)*
<!-- Sent from Reviewable.io -->

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
  +0.0%    +227 [None]                                                                           +4.65Ki  +0.1%
  +3.3%     +48 src/core/lib/iomgr/iomgr.cc                                                          +48  +3.3%
      [NEW]     +41 grpc_iomgr_count_objects_for_testing                                                 +41  [NEW]
       +17%      +7 [Unmapped]                                                                            +7   +17%
  +1.4%     +32 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver.cc       +32  +1.4%
      [NEW]    +380 grpc_ares_ev_driver_create_locked                                                   +380  [NEW]
       +24%     +12 [Unmapped]                                                                           +12   +24%
      [NEW]      +9 grpc_ares_ev_driver_on_queries_complete_locked                                        +9  [NEW]
      +4.7%      +8 grpc_ares_ev_driver_unref                                                             +8  +4.7%
  +3.6%     +13 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver_posi     +13  +3.6%
       +13%     +13 grpc_core::GrpcPolledFdPosix::~GrpcPolledFdPosix                                     +13   +13%
      +4.0%      +4 grpc_core::NewGrpcPolledFdLocked                                                      +4  +4.0%

  +0.0%    +320 TOTAL                                                                            +4.74Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

markdroth commented:
Combining the objects in a separate PR sounds fine.  Thanks!

---

Reviewed 1 of 1 files at r9.
Review status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @dgquintas, @yihuazhang, @sreecha, and @AspirinSJL)

---



*Comments from [Reviewable](https://reviewable.io/reviews/grpc/grpc/15828#-:-LFx1KRv9qR9hLSROrj0:b-ecorh3)*
<!-- Sent from Reviewable.io -->

apolcyn commented:
squashed down commits

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
  +0.0%    +259 [None]                                                                           +4.65Ki  +0.1%
  +3.3%     +48 src/core/lib/iomgr/iomgr.cc                                                          +48  +3.3%
      [NEW]     +41 grpc_iomgr_count_objects_for_testing                                                 +41  [NEW]
       +17%      +7 [Unmapped]                                                                            +7   +17%
  +1.4%     +32 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver.cc       +32  +1.4%
      [NEW]    +380 grpc_ares_ev_driver_create_locked                                                   +380  [NEW]
       +24%     +12 [Unmapped]                                                                           +12   +24%
      [NEW]      +9 grpc_ares_ev_driver_on_queries_complete_locked                                        +9  [NEW]
      +4.7%      +8 grpc_ares_ev_driver_unref                                                             +8  +4.7%
  +3.6%     +13 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver_posi     +13  +3.6%
       +13%     +13 grpc_core::GrpcPolledFdPosix::~GrpcPolledFdPosix                                     +13   +13%
      +4.0%      +4 grpc_core::NewGrpcPolledFdLocked                                                      +4  +4.0%

  +0.0%    +352 TOTAL                                                                            +4.74Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

apolcyn commented:
bazel test failures: https://github.com/grpc/grpc/issues/14240
